waldo/host/src/bin/prove.rs

```diff
@@ -0,0 +1,125 @@
+use std::error::Error;
+use std::fs;
+use std::path::PathBuf;
+
+use clap::Parser;
+use image::io::Reader as ImageReader;
+use image::{imageops, GenericImageView};
+use risc0_zkvm::prove::{Prover, ProverOpts};
+use risc0_zkvm::serde;
+use waldo_core::image::{ImageMerkleTree, IMAGE_CHUNK_SIZE};
+use waldo_core::merkle::VECTOR_ORACLE_CHANNEL;
+use waldo_core::PrivateInput;
+use waldo_methods::{IMAGE_CROP_ID, IMAGE_CROP_PATH};
+
+#[derive(Parser, Debug)]
+#[clap(author, version, about, long_about = None)]
+struct Args {
+    /// Input file path to the full Where's Waldo image.
+    #[clap(short = 'i', long, value_parser, value_hint = clap::ValueHint::FilePath)]
+    image: PathBuf,
+
+    /// X coordinate, in pixels from the top-left corner, of Waldo.
+    #[clap(short = 'x', long, value_parser)]
+    waldo_x: u32,
+
+    /// Y coordinate, in pixels from the top-left corner, of Waldo.
+    #[clap(short = 'y', long, value_parser)]
+    waldo_y: u32,
+
+    /// Width, in pixels, of the cutout for Waldo.
+    #[clap(short = 'w', long, value_parser)]
+    waldo_width: u32,
+
+    /// Height, in pixels, of the cutout for Waldo.
+    #[clap(short = 'h', long, value_parser)]
+    waldo_height: u32,
+
+    /// Optional input file path to an image mask to apply to Waldo.
+    /// Grayscale pixel values will be subtracted from the cropped image of Waldo such that a black
+    /// pixel in the mask will result in the cooresponding image pixel being blacked out.
+    /// Must be the same dimensions, in pixels, as the cut out x and y.
+    #[clap(short = 'm', long, value_parser, value_hint = clap::ValueHint::FilePath)]
+    mask: Option<PathBuf>,
+
+    /// Output file path to save the receipt. Note that the receipt contains the cutout of waldo.
+    #[clap(short = 'r', long, value_parser, default_value = "./receipt.bin", value_hint = clap::ValueHint::FilePath)]
+    receipt: PathBuf,
+}
+
+fn main() -> Result<(), Box<dyn Error>> {
+    let args = Args::parse();
+
+    // Read the image from disk.
+    let img = ImageReader::open(&args.image)?.decode()?;
+    println!(
+        "Read image at {} with size: {} x {}",
+        &args.image.display(),
+        img.width(),
+        img.height()
+    );
+
+    let crop_location = (args.waldo_x, args.waldo_y);
+    let crop_dimensions = (args.waldo_width, args.waldo_height);
+
+    // Read the image mask from disk, if provided.
```

Comment:
Doesn't the image mask let me fake a picture of waldo?  Can't I find an all black part of the image and then make the mask look like waldo?  Perhaps if we make the mask required to be binary or put the mask hash into the journal (or both).  Or skip the mask I suppose.

Reply:
I did chose grayscale for this reason, but I went ahead and simplified the masking in ed764fba36f5a6cccb285925c6c0d8f261003747 by making the values interpreted as booleans (i.e. `0` is "mask out" and everything else leaves the pixel unmodified). I thought about including the mask in the journal, but decided against it to keep what needs to be verified minimal. (My thought being that for the fact that the mask is in the journal to have any significance, it would have to be something the user checks.)